File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.cpp
#include <string>
#include <vector>

#include <unordered_map>
#include <algorithm>

using namespace std;

// second 기준 내림차순
bool compare(const pair<string, int>& p1, const pair<string, int>& p2) {
    return p1.second > p2.second;
}

void comb(string prefix, string postfix, int length, int depth, unordered_map<string, int>& count) {
    // 현재 자기 depth를 바탕으로 for문을 어느 범위로 순회할지 결정해야 함.
    if (depth == length) {
        for(char ch: postfix) {
            count[prefix + ch]++;
        }
    }
    else {
        int l = postfix.length();
        for(int i=0; i<postfix.length(); i++) {
            comb(prefix + postfix[i], postfix.substr(i+1,l-i-1), length, depth+1, count);
        }
    }
    
}

vector<string> solution(vector<string> orders, vector<int> course) {
    // 구성 메뉴 개수 - 조합별 카운트
    unordered_map<int, unordered_map<string, int>> counts;
    
    // 1. 각 손님의 주문을 사전 순 정렬한다.
    for(string& order: orders) {
        sort(order.begin(), order.end());
    }
    
    // 2. 각 손님의 주문 조합에서 나올 수 있는 모든 조합을 생성해 카운트한다.
    for(string order: orders) {
        for(int c: course) {
            if (c <= order.length()) comb("", order, c, 1, counts[c]);
        }
    }
    
    // 3. course에 나온 길이별 조합 중에서 가장 카운트가 많은 조합을 선택한다.
    vector<string> answer;
    for(int c: course) {
        if(counts[c].empty()) continue;
        
        // 길이가 c인 메뉴 조합들(counts[c] 원소들)을 vector에 저장 후 sort
        vector<pair<string, int>> combs(counts[c].begin(), counts[c].end());
        sort(combs.begin(), combs.end(), compare);
        int _maxL = combs[0].second;
        if (_maxL < 2) continue; // 가장 많이 선택한 메뉴가 1인 경우 추가하지 않음
        
        // 해당 길이의 메뉴조합 중 가장 많이 선택한 메뉴 조합을 answer에 push
        for(auto &p: combs) {
            if (p.second == _maxL) answer.push_back(p.first);
            else break;
        }
        
    }
    
    sort(answer.begin(), answer.end()); // 마지막으로 answer에 추가된 메뉴들을 사전 순 정렬
    return answer;
}

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;

class Solution {
    
    static Map<String, Integer> counts = new HashMap<>();
    
    static void comb(String prefix, String postfix, int length) {
        // 목표한 길이에 도달하면 map에 카운트
        if (prefix.length() == length) {
            if (!counts.containsKey(prefix)) counts.put(prefix, 1);
            else counts.put(prefix, counts.get(prefix)+1);
        }
        else {
            for(int i=0; i<postfix.length(); i++) {
                comb(prefix + postfix.charAt(i), postfix.substring(i+1), length);
            }
        }
    }
    
    static String sortStr(String s) {
        char[] buf = s.toCharArray();
        Arrays.sort(buf);
        return new String(buf);
    }
    
    public String[] solution(String[] orders, int[] course) {
        // 1. 각 order를 사전 순으로 정렬한다.
        List<String> orderL = new ArrayList<>();
        for(String s: orders)
            orderL.add(sortStr(s));
        
        // 2. 각 order를 통해 만들 수 있는 조합을 문자열로 표현하고, 그 빈도를 Map에 저장한다. (static Map)
        List<String> ans = new ArrayList<>(); // 답
        
        List<String> ranking; // 문자열로 표현된 각 조합을 빈도 순으로 정렬할 리스트
        for(int length: course) {
            
            // 각 order에 대해 length 길이로가능한 조합들을 map에 저장
            for(String o: orderL) {
                if (o.length() < length) continue; // 길이가 짧은 것은 건너 뜀
                comb("", o, length);
            }
            
            // counts 값 순으로 정렬
            ranking = new ArrayList<>(counts.keySet());
            ranking.sort(new Comparator<String>() {
                @Override
                public int compare(String s1, String s2) {
                    int n1 = counts.get(s1);
                    int n2 = counts.get(s2);
                    return Integer.compare(n2, n1); // 내림차순
                }
            });
            
            if (ranking.size() == 0) continue; // 해당 length에 해당하는 조합이 아예 없을 수도 있음
            int top = counts.get(ranking.get(0));
            if (top < 2) continue; // 2번 이상 나온 조합에 대해서만 추가
            for(String s: ranking) {
                if (counts.get(s) == top) ans.add(s);
                else break;
            }
            
            counts.clear(); // 다음 length에 대해 카운트하기 위해 map 초기화
        }
        
        String[] answer = new String[ans.size()];
        int seq = 0;
        for(String s: ans)
            answer[seq++] = s;
        
        Arrays.sort(answer);
        return answer;
    }
}